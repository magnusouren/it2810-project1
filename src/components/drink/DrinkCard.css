@import url('https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@20..48,100..700,0..1,-50..200');

.drinkContainer {
  max-width: 600px;
  margin: 0 auto;
  padding: 0 30px;
  display: flex;
  flex-flow: row wrap;
  overflow-x: hidden;
}

.drinkContainer .info,
.drinkContainer .ingredients,
.drinkContainer .instructions {
  font-family: Lato, sans-serif;
  font-weight: 300;
}

.drinkContainer .ingredients,
.drinkContainer .instructions {
  margin-top: 40px;
  padding: 0;
  flex-basis: 50%;
}

.drinkContainer .sectionHeader {
  font-family: raleway, Lato, sans-serif;
  font-weight: 500;
  margin-bottom: 10px;
}

/* **** Return Button **** */
.drinkContainer .returnButtonContainer .returnButton {
  font-size: 2em;
  color: #fbc4ab;
  margin-right: 0.5em;
  font-weight: 300;
  position: absolute;
  top: 36px;
  left: 320px;
}

.drinkContainer .returnButtonContainer .returnButton:hover {
  color: #f08080;
}

/* **** Header **** */
.drinkContainer .drinkHeader {
  height: 110px;
  max-height: 110px;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-basis: 100%;
}

.drinkContainer h1 {
  text-align: center;
  padding: 0;
  font-size: 2em;
  font-family: Raleway, Lato, sans-serif;
  font-weight: 500;
}

/* **** Image **** */
.drinkContainer .imageContainer {
  display: flex;
  justify-content: center;
  align-items: center;
  margin: 0 auto;
  flex-basis: 100%;
}

.drinkContainer .imageContainer img {
  height: 300px;
  overflow-x: hidden;
  border-radius: 8px;
  box-shadow:
    0 4px 8px 0 rgb(0 0 0 / 20%),
    0 6px 20px 0 rgb(0 0 0 / 19%);
}

/* **** Info **** */

.drinkContainer .info {
  margin: 40px auto 0;
  padding: 0;
  width: 100%;
  display: flex;
  flex-flow: row wrap;
  justify-content: space-between;
  align-items: center;
  color: gray;
}

.drinkContainer .info p::before {
  font-family: 'Material Symbols Outlined';
  font-size: 1.5em;
  color: #fbc4ab;
  margin-right: 6px;
  font-weight: 300;
}

.drinkContainer .info #drinkCategory::before,
.drinkContainer .info #drinkGlass::before,
.drinkContainer .info #drinkAlcoholic::before {
  vertical-align: -20%;
}

.drinkContainer .info #drinkCategory::before {
  content: '\e574';
}

.drinkContainer .info #drinkGlass::before {
  content: '\e540';
}

.drinkContainer .info #drinkAlcoholic::before {
  content: '\ea60';
}

/* **** Ingredients **** */

.drinkContainer .ingredientsList li {
  display: flex;
  align-items: center;
  margin-right: 10px;
  margin-bottom: 10px;
}

.drinkContainer .ingredientsList li div:first-child {
  min-width: 120px;
}

@media screen and (width <= 924px) {
  .drinkContainer .info p {
    margin-right: 20px;
  }

  .drinkContainer .info p:last-child {
    margin-right: 0;
  }

  .drinkContainer .ingredients,
  .drinkContainer .instructions {
    margin-top: 20px;
    flex-basis: 100%;
  }

  .drinkContainer .instructions p {
    max-width: 70%;
  }
}

@media screen and (width <= 790px) {
  .drinkContainer .instructions p {
    max-width: 100%;
  }
}

@media screen and (width <= 680px) {
  .drinkContainer .info {
    flex-direction: column;
    align-items: start;
  }
}
