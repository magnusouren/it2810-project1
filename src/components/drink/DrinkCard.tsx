import axios from 'axios';
import React, { FC, useEffect, useState } from 'react';

interface Ingredient {
  ingredient: string;
  measure: string;
}

interface Drink {
  strDrink: string;
  ingredients: Array<Ingredient>;
  strGlass: string | null;
  strInstructions: string;
  strDrinkThumb: string;
  strVideo: string | null;
  strCategory: string;
  strAlcoholic: string;
}

export const DrinkCard: FC = () => {
  const [drink, setDrink] = useState<Drink | null>(null);

  useEffect(() => {
    console.log('Fetching drink...');
    // Fetch a random drink from the API
    axios
      .get('https://www.thecocktaildb.com/api/json/v1/1/search.php?s=margarita')
      .then((response) => {
        const drinkData = response.data.drinks[0];
        const ingredients: Array<Ingredient> = [];

        // Extract ingredients
        for (let i = 1; i <= 15; i++) {
          const ingredient: string = drinkData[`strIngredient${i}`];
          const measure: string = drinkData[`strMeasure${i}`];
          if (ingredient) {
            ingredients.push({ ingredient, measure });
          } else {
            break;
          }
        }

        drinkData.ingredients = ingredients;
        setDrink(drinkData);
      })
      .catch((error) => {
        console.error('Error fetching drink:', error);
      });
  }, []);

  if (!drink) {
    return <div>Loading...</div>;
  }

  return (
    <div className='drinkContainer'>
      <h2>{drink.strDrink}</h2>
      <div className='imageContainer'>
        <img src={drink.strDrinkThumb} alt={drink.strDrink + 'image'} />
      </div>
      <p>{drink.strVideo}</p>
      <div className='misc'>
        <p>{drink.strCategory}</p>
        <p>{drink.strGlass}</p>
        <p>{drink.strAlcoholic}</p>
      </div>
      <div>
        <h3>Instructions</h3>
        <p>{drink.strInstructions}</p>
      </div>
      <ul className='ingredients'>
        {drink.ingredients.map((ingredient, index) => (
          <li key={index}>
            <div className='measure'>{ingredient.measure == null ? '' : ingredient.measure}</div>
            <div className='ingredient'>{ingredient.ingredient}</div>
          </li>
        ))}
      </ul>
    </div>
  );
};
