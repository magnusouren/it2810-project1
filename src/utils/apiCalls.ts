import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
