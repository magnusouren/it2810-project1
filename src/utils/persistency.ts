import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
