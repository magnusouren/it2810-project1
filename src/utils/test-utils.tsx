import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
