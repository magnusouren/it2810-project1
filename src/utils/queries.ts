import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
