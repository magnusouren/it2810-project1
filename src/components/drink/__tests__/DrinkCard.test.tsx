import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { DrinkCard } from '../DrinkCard';

/* Minimal API return format */
const testDrink = {
  drinkId: '11118',
  strDrink: 'Super Drink',
  ingredients: [],
  strGlass: 'Highball',
  strInstructions: 'Instructions for drink',
  strDrinkThumb: 'https://localhost:3000/drink/11118.jpg',
  strCategory: 'Category',
  strAlcoholic: 'Alcoholic',
  strIngredient1: 'Ingredient 1',
  strMeasure1: 'Measure 1',
  strIngredient2: 'Ingredient 2',
  strMeasure2: 'Measure 2',
};

const tree = (id: string) => {
  return (
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>
  );
};

describe('DrinkCard', () => {
  it('Renders with proper params', async () => {
    render(tree(testDrink.drinkId));

    await waitFor(() => {
      expect(screen.getByText(testDrink.strDrink)).toBeDefined();
      expect(screen.getByText(testDrink.strGlass)).toBeDefined();
      expect(screen.getByText(testDrink.strInstructions)).toBeDefined();
      expect(screen.getByText(testDrink.strCategory)).toBeDefined();
      expect(screen.getByText(testDrink.strAlcoholic)).toBeDefined();
      expect(screen.getByText(testDrink.strIngredient1)).toBeDefined();
      expect(screen.getByText(testDrink.strMeasure1)).toBeDefined();
      expect(screen.getByText(testDrink.strIngredient2)).toBeDefined();
      expect(screen.getByText(testDrink.strMeasure2)).toBeDefined();
    });
  });

  it('Should match snapshot', async () => {
    const { container } = render(tree(testDrink.drinkId));

    /* Expecting Drink name to show to make sure
    the component is finished loading before comparing snapshots */
    await waitFor(() => {
      expect(screen.getByText(testDrink.strDrink)).toBeDefined();
    });
    expect(container).toMatchSnapshot();
  });

  it('Should show loading state when data is loading', () => {
    renderWithRouterAndQueryClient(<DrinkCard />);
    expect(screen.getByText(/Loading/i)).toBeDefined();
  });

  it.skip('Should show that no drinks were found if none are returned', async () => {
    render(tree('0'));

    await waitFor(() => {
      expect(screen.getByText(/No drink was found.../i)).toBeDefined();
    });
  });
});
