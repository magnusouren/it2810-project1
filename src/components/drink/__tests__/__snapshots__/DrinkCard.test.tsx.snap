// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrinkCard > Should match snapshot 1`] = `
<div>
  <div
    class="drink-container"
  >
    <div
      class="drink-header"
    >
      <div
        class="return-button-container"
      >
        <button
          class="return-button"
        >
          &lt;-
        </button>
      </div>
      <h1>
        Super Drink
      </h1>
      <div
        class="favorite-button-container"
      >
        <div
          class="favorite-toggle"
          data-testid="favorite"
        >
          &lt;3
        </div>
      </div>
    </div>
    <div
      class="image-container"
    >
      <img
        alt="Super Drinkimage"
        src="https://localhost:3000/drink/11118.jpg"
      />
    </div>
    <div
      class="info"
    >
      <p
        id="drink-category"
      >
        Category
      </p>
      <p
        id="drink-glass"
      >
        Highball
      </p>
      <p
        id="drink-alcoholic"
      >
        Alcoholic
      </p>
    </div>
    <div
      class="ingredients"
    >
      <h3
        class="section-header"
      >
        Ingredients
      </h3>
      <ul
        class="ingredients-list"
      >
        <li>
          <div
            class="measure"
          >
            Measure 1
          </div>
          <div
            class="ingredient"
          >
            Ingredient 1
          </div>
        </li>
        <li>
          <div
            class="measure"
          >
            Measure 2
          </div>
          <div
            class="ingredient"
          >
            Ingredient 2
          </div>
        </li>
      </ul>
    </div>
    <div
      class="instructions"
    >
      <h3
        class="section-header"
      >
        Instructions
      </h3>
      <p>
        Instructions for drink
      </p>
    </div>
  </div>
</div>
`;
